Flaky TestDistributor_PushInstanceLimits

The `TestDistributor_PushInstanceLimits` is flaky since #398:

```
--- FAIL: TestDistributor_PushInstanceLimits (0.39s)
    --- FAIL: TestDistributor_PushInstanceLimits/below_inflight_limit (0.01s)
        distributor_test.go:618: 
            	Error Trace:	distributor_test.go:618
            	Error:      	Received unexpected error:
            	            	
            	            	metric output does not match expectation; want:
            	            	
            	            	# HELP cortex_distributor_inflight_push_requests Current number of inflight push requests in distributor.
            	            	# TYPE cortex_distributor_inflight_push_requests gauge
            	            	cortex_distributor_inflight_push_requests 100
            	            	# HELP cortex_distributor_instance_limits Instance limits used by this distributor.
            	            	# TYPE cortex_distributor_instance_limits gauge
            	            	cortex_distributor_instance_limits{limit="max_inflight_push_requests"} 101
            	            	cortex_distributor_instance_limits{limit="max_ingestion_rate"} 0
            	            	
            	            	got:
            	            	
            	            	# HELP cortex_distributor_inflight_push_requests Current number of inflight push requests in distributor.
            	            	# TYPE cortex_distributor_inflight_push_requests gauge
            	            	cortex_distributor_inflight_push_requests 101
            	            	# HELP cortex_distributor_instance_limits Instance limits used by this distributor.
            	            	# TYPE cortex_distributor_instance_limits gauge
            	            	cortex_distributor_instance_limits{limit="max_inflight_push_requests"} 101
            	            	cortex_distributor_instance_limits{limit="max_ingestion_rate"} 0
            	Test:       	TestDistributor_PushInstanceLimits/below_inflight_limit
FAIL
FAIL	github.com/grafana/mimir/pkg/distributor	42.341s
```

--- a/pkg/distributor/distributor_test.go
+++ b/pkg/distributor/distributor_test.go
@@ -615,7 +615,12 @@ func TestDistributor_PushInstanceLimits(t *testing.T) {
 				d.ingestionRate.Tick()
 
 				if testData.expectedMetrics != "" {
-					assert.NoError(t, testutil.GatherAndCompare(regs[0], strings.NewReader(testData.expectedMetrics), testData.metricNames...))
+					// The number of inflight requests is decreased asynchronously once the request to the latest
+					// ingester is completed too. To avoid flaky tests, we poll the metrics because what we expect
+					// is that metrics reconcile to the expected ones.
+					test.Poll(t, 3*time.Second, nil, func() interface{} {
+						return testutil.GatherAndCompare(regs[0], strings.NewReader(testData.expectedMetrics), testData.metricNames...)
+					})
 				}
 			}
 		})

